core_lib/src/util/util.cpp

```diff
@@ -126,3 +128,59 @@ QString uniqueString(int len)
     s[len] = 0;
     return QString::fromUtf8(s);
 }
+
+QString validateDataPath(QString filePath, QString dataDirPath)
+{
+    // Make sure src path is relative
+    if (!QFileInfo(filePath).isRelative()) return QString();
+
+    QFileInfo fi(dataDirPath, filePath);
+    // Recursively resolve symlinks
+    QString canonicalPath = fi.canonicalFilePath();
+
+    // Iterate over parent directories of the file path to see if one of them equals the data directory
+    QDir dataDir(dataDirPath);
+    if (dataDir.exists()) dataDir.setPath(dataDir.canonicalPath());
```

Comment:
I don't really get what this line does? is it for resolving the symlinks?
Also could you add curly brackets?

Reply:
Yes. The canonical path is effectively the absolute path with all symlinks and relative components ('.' and '..') resolved. This particular line is necessary to deal with the case when one or more of the directories in the data dir's path is a symlink. For example, imagine if the data dir for a project is in `/tmp/Pencil2D/Project_Y2xD_012345/data`, and `/tmp` is actually a symlink to `/custom_temp_dir`. In this case, a vector file defined in main.xml could be `002.001.vec`, which when that file is fully resolved below, it will result in a path of `/custom_temp_dir/Pencil2D/Project_Y2xD_012345/data/002.001.vec`. This will not match the data dir value of `/tmp/Pencil2D/Project_Y2xD_012345/data` unless the symlinks are also resolved for the data dir path.

Reply:
good thank you for explaining this.